swarm/storage/localstore/mode_put.go

```diff
@@ -0,0 +1,165 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package localstore
+
+import (
+	"github.com/ethereum/go-ethereum/swarm/shed"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+// ModePut enumerates different Putter modes.
+type ModePut int
+
+// Putter modes.
+const (
+	// ModePutRequest: when a chunk is received as a result of retrieve request and delivery, it is put only in
+	ModePutRequest ModePut = iota
+	// ModePutSync: when a chunk is received via syncing in it is put in
+	ModePutSync
+	// ModePutUpload: when a chunk is created by local upload it is put in
+	ModePutUpload
+)
+
+// Putter provides Put method to store Chunks
+// to database.
+type Putter struct {
+	db   *DB
+	mode ModePut
+}
+
+// NewPutter returns a new Putter on database
+// with a specific Mode.
+func (db *DB) NewPutter(mode ModePut) *Putter {
+	return &Putter{
+		mode: mode,
+		db:   db,
+	}
+}
+
+// Put stores the Chunk to database and depending
+// on the Putter mode, it updates required indexes.
+func (p *Putter) Put(ch storage.Chunk) (err error) {
+	return p.db.put(p.mode, chunkToItem(ch))
+}
+
+// put stores Item to database and updates other
+// indexes. It acquires lockAddr to protect two calls
+// of this function for the same address in parallel.
+// Item fields Address and Data must not be
+// with their nil values.
+func (db *DB) put(mode ModePut, item shed.Item) (err error) {
+	// protect parallel updates
+	if db.useGlobalLock {
+		db.globalMu.Lock()
+		defer db.globalMu.Unlock()
+	} else {
+		unlock, err := db.lockAddr(item.Address)
+		if err != nil {
+			return err
+		}
+		defer unlock()
+	}
+
+	batch := new(leveldb.Batch)
+
+	// variables that provide information for operations
+	// to be done after write batch function successfully executes
+	var gcSizeChange int64   // number to add or subtract from gcSize
+	var triggerPullFeed bool // signal pull feed subscriptions to iterate
+	var triggerPushFeed bool // signal push feed subscriptions to iterate
+
+	switch mode {
+	case ModePutRequest:
+		// put to indexes: retrieve, gc; it does not enter the syncpool
+
+		// check if the chunk already is in the database
+		// as gc index is updated
+		i, err := db.retrievalAccessIndex.Get(item)
+		switch err {
+		case nil:
+			item.AccessTimestamp = i.AccessTimestamp
+		case leveldb.ErrNotFound:
+			// no chunk accesses
+		default:
+			return err
+		}
+		i, err = db.retrievalDataIndex.Get(item)
+		switch err {
+		case nil:
+			item.StoreTimestamp = i.StoreTimestamp
+		case leveldb.ErrNotFound:
+			// no chunk accesses
+		default:
+			return err
+		}
+		if item.AccessTimestamp != 0 {
+			// delete current entry from the gc index
+			db.gcIndex.DeleteInBatch(batch, item)
```

Comment:
looks like GC logic is dispersed over different places...

Reply:
Any ideas how to organize it better? My current one is to use global lock, which would simplify gc counting. But, this part of the code that you are referring, just needs to remove the old item in gc index and put the new one with new access timestamp. Any ideas how to simplify it?